### PR TITLE
make convoutional extensions more maintainable

### PR DESCRIPTION
**Pull request by DerThorsten**
This pr tries to cleanup the file `convolutional.py`.
Most of the code is very redundant and hard to maintain.

Here I propose the following:
 - Instead of having 'handwritten classes for all activation and dimensions lets create them dynamically
- using `functools.partialmethod`on `ConvActivation.__init__` does this trick
- to make this more convenient I added a method called `register_partial_cls` in the spirit of   `functools.partialmethod`

Consider this example where we have a base class `Conv` and want to create 1D and 2D classes
```python
    class Conv(object):
        def __init__(self, dim, activation, stride=1):
            pass
```
Currently we are doing smth like:
```python
    class Conv1D(Conv):
        def __init__(self, activation, stride=1):
            super().__init__(activation=activation,
                dim=1,
                stride=stride)

    class Conv2D(Conv):
        def __init__(self, activation, stride=1):
            super().__init__(activation=activation,
                dim=2,
                stride=stride)
```

Using the proposed functionality one can write:
``` python
    for dim in [1,2]:
        register_partial_cls(Conv, 'Conv{}D'.format(dim), __name_, globals(), dim=2)
```
In both cases, the usage is:
```python
    b = Conv1D(activation='ELU')
    b = Conv2D(activation='ELU')
```

There is one downside: `functools.partial` does not remove the keyword which is binded, but rather introduces a (new) default value for the keyword/argument. This means that the for instance this code would work

``` python
    for dim in [1,2]:
        register_partial_cls(Conv, 'Conv{}D'.format(dim), globals(), dim=2)
    
    b = Conv1D(activation='ELU, dim=2)
```
But I think this could be fixed by re-implementing smth like functools partial, but this is not yet part of this pr (I envision smth. where one can specific which keywords are binded and removed, and for which keywords just the default values are updated) 

I would argue, that even with the above-mentioned downside / caveat , this PR is still worthwhile and can be merged as is since the drastic increase of maintainability / reduced lines of code

With a one additional helper function we can create allmost all convenient classes (besides
the batchnorm classes, since they are a bit different ) in **~40 lines of code** (instead of ~400)

```python
....
from ...utils.partial_cls import register_partial_cls
register_partial_cls_here = functools.partial(register_partial_cls, module=__name__, module_dict=globals())

def _register_conv_cls(prefix, **kwargs):
    # simple conv activation
    activations = ["ReLU", "ELU", "Sigmoid", "SELU", ""]
    init_map = {
        "ReLU": KaimingNormalWeightsZeroBias,
        "SELU": SELUWeightsZeroBias
    }
    for activation_str in activations:
        cls_name = cls_name = "{}{}".format(prefix,activation_str)
        initialization_cls = init_map.get(activation_str, OrthogonalWeightsZeroBias)
        if activation_str == "":
            activation = None
        elif activation_str == "SELU":
            if hasattr(nn, "SELU"):
                # Pytorch 0.2: Use built in SELU
                activation = nn.SELU(inplace=True)
            else:
                # Pytorch < 0.1.12: Use handmade SELU
                activation = SELU()
        else:
            activation = activation_str
        register_partial_cls_here(ConvActivation, cls_name,
            activation=activation,
            **kwargs)
        for dim in [1, 2, 3]:
            cls_name = "{}{}{}D".format(prefix,activation_str, dim)
            print("cls_name", cls_name)
            register_partial_cls_here(ConvActivation, cls_name,
                dim=dim, 
                activation=activation,
                initialization=initialization_cls(),
                **kwargs)

_register_conv_cls("Conv")
_register_conv_cls("ValidConv",  valid_conv=True)
_register_conv_cls("Deconv", deconv=True, kernel_size=2, stride=2)
_register_conv_cls("StridedConv",  stride=2)
_register_conv_cls("DilatedConv",  dilation=2)

del _register_conv_cls
```

### CODE REVIEW

**Comment by DerThorsten** [comment]
The pickling of the dynamic classes still fails, but I think I'll have that fixed soon

**Comment by DerThorsten** [comment]
Maybe this is a showstopper :( seems like it is hard to pickle dynamic classes
```
src/inferno/tests/test_inferno.py::TestInferno::test_training_cpu
  /home/thorsten/miniconda3/envs/pydev37/lib/python3.7/site-packages/torch/serialization.py:256: UserWarning: Couldn't retrieve source code for container of type Conv2D. It won't be checked for correctness upon loading.
    "type " + obj.__name__ + ". It won't be checked "

-- Docs: https://docs.pytest.org/en/latest/warnings.html
```

**Comment by DerThorsten** [comment]
update:
with a custom version of partial I wrote one can now specify which arguments are fixed, and for which we just want to update the default value, with that

``` python
    for dim in [1,2]:
        register_partial_cls(Conv, 'Conv{}D'.format(dim), globals(), dim=2)
    
    b = Conv1D(activation='ELU, dim=2)
```
will raise an `TypeError`

The code which registers the conv classes is now called like this:

```python

_register_conv_cls("Conv")
_register_conv_cls("ValidConv",  fixed=dict(valid_conv=True))
_register_conv_cls("Deconv", fixed=dict(deconv=True), default=dict(kernel_size=2, stride=2))
_register_conv_cls("StridedConv",  fixed=dict(stride=2))
_register_conv_cls("DilatedConv",  fixed=dict(dilation=2))
```

File: inferno/extensions/layers/convolutional.py [hunk 1]
```diff
@@ -1,77 +1,125 @@
...
-           'Conv2D', 'Conv3D',
-           'BNReLUConv2D', 'BNReLUConv3D',
-           'BNReLUDepthwiseConv2D',
-           'ConvSELU2D', 'ConvSELU3D',
-           'ConvReLU2D', 'ConvReLU3D',
-           'BNReLUDilatedConv2D', 'DilatedConv2D',
-           'GlobalConv2D']
+from ...utils.partial_cls import register_partial_cls
+
+__all__ = [
```
Comment (constantinpape) [inline inferno/extensions/layers/convolutional.py]: Could we autogenerate this as well, by appending the names to `__all__` in `_register_conv_cls`?

Comment (DerThorsten) [reply inferno/extensions/layers/convolutional.py]: yes, we could append to `__all__` or just remove the `__all__` in which case python basicaly generates this for us.

**Comment by DerThorsten** [comment]
Jeah there are still a few minor things to fix, I  think they should be done by tomorrow

**Comment by DerThorsten** [comment]
Looks like the pytorch>=1.0 test is passing, Are there plans to drop 0.4 support?

**Comment by constantinpape** [comment]
> Looks like the pytorch>=1.0 test is passing, Are there plans to drop 0.4 support?

Would be fine with me. Any thoughts @nasimrahaman?

**Comment by DerThorsten** [comment]
Still struggling to understand why https://github.com/inferno-pytorch/inferno/pull/189/commits/197dc9d7275827c828a52a81221936e1f07af1aa passed and the rest not

**Comment by nasimrahaman** [comment]
Sure, I think it’s safe to drop 0.4 (current version is 1.2).

**Comment by DerThorsten** [comment]
ok, then let me polish this and clean it up properly

**Comment by DerThorsten** [comment]
I think this good to go now.
Not only is the numbers of lines drastically reduced, but also a lot of new flavors  are added
like `BNReLUStridedConv1D`.

One think where I was not sure is how to name the `ND` classes (those which have `dim` as kw-arg).
The options are `Conv` or `ConvND`. I think we need to do `ConvND` for the following reasons:
 - I also implemented a `ND` batchnorm, 
 - in pytorch there is already a class `nn.Batchnorm` which is not  what we mean`BatchNorrmND` since the `dim` kw is missing. 
- it would be confusing if `inferno.Batchnorm` would behave different from `nn.Batchnorm` => it needs to be `inferneo.BatchnormND` => therefore also `ConvND`, `DeconvND` etc.

Side remark, pytorch uses `2d` we use `2D` which is abit ugly.

**Comment by DerThorsten** [comment]
I'll merge this now,
since we require pytorch >=1.0 with this pr we should bump the minor version soonish.
But first we should remove all old stuf like the selfmade `SELU` for pytorch < 0.2 and so on
